src/dot.ts

```diff
@@ -125,20 +139,20 @@ export class Dot extends Modifier {
     return this;
   }
 
-  setDotShiftY(y) {
+  setDotShiftY(y: number): this {
     this.dot_shiftY = y;
     return this;
   }
 
-  draw() {
-    this.checkContext();
+  draw(): void {
+    const ctx = this.checkContext();
     this.setRendered();
 
-    if (!this.note || this.index === null) {
+    if (!this.note || this.index === undefined) {
       throw new Vex.RERR('NoAttachedNote', "Can't draw dot without a note and index.");
     }
 
-    const lineSpace = this.note.stave.options.spacing_between_lines_px;
+    const lineSpace = this.note.getStave()?.getOptions().spacing_between_lines_px ?? 0;
```

Comment:
This is a potential silent failure, let's use `getStave()!` instead (or explicitly check for stave above.)

Reply:
`getStave()!` is also silent you get in the end `undefined` and that's it. In **migration/tmisignote** we got even `this.timeSig!.line!` when `line` was mo known attribute.

Reply:
Done

Reply:
Is it really silent, I thought the `!` causes an exception if the value is undefined, no?

Reply:
No compilation failure and no runtime failure. This is what we observed in **migration/timesignote** with `this.timeSig!.line!`
@ronyeh can you confirm?
I believe this is whay this a deprecated feature in eslint.

Reply:
Oh, no, I completely misunderstood the `!` operator -- I thought it would be a runtime assertion, but it doesn't emit any code at all. (The term "non-null assertion operator" make me think it was an actual assertion.)

In that case, yes 100% this should be completely avoided, and we should use explicit checks like `if (something not defined) { throw exception };`

Apologies for this, can we re add the es-lint rule and replace all the `!` operators with explicit checks (where reasonable), maybe in a separate issue/PR?

Reply:
I will do that once we closed the open PRs. I will open no new one and then I will remove the allowance and go though all the cases. I also understood it wrong at the beginning, it is very misleading!